src/ansi.rs

```diff
@@ -0,0 +1,1849 @@
+#[derive(Default)]
+pub struct NullTimeProvider {}
+
+impl TimeProvider for NullTimeProvider {
+    fn now(&self) -> u128 {
+        0
+    }
+}
+
+#[derive(Debug)]
+#[cfg(feature = "alloc")]
+struct SyncState<T: TimeProvider> {
+    /// Expiration time (in milliseconds since unix epoch) of the synchronized update.
+    timeout: Option<u128>,
+
+    in_sync: bool,
+
+    /// Sync DCS waiting for termination sequence.
+    pending_dcs: Option<Dcs>,
+
+    /// Bytes read during the synchronized update.
+    buffer: Vec<u8>,
+
+    /// Time provider
+    time_provider: T,
+}
+
+#[cfg(feature = "alloc")]
+impl<T: TimeProvider> SyncState<T> {
+    fn new(time_provider: T) -> SyncState<T> {
+        Self {
+            buffer: Vec::with_capacity(SYNC_BUFFER_SIZE),
+            pending_dcs: None,
+            timeout: None,
+            in_sync: false,
+            time_provider,
+        }
+    }
+}
+
+/// Pending DCS sequence.
+#[derive(Debug)]
+#[cfg(feature = "alloc")]
+enum Dcs {
+    /// Begin of the synchronized update.
+    SyncStart,
+
+    /// End of the synchronized update.
+    SyncEnd,
+}
+
+/// The processor wraps a `Parser` to ultimately call methods on a Handler.
+pub struct Processor<T: TimeProvider> {
+    state: ProcessorState<T>,
+    parser: Parser,
+}
+
+impl Processor<NullTimeProvider> {
+    #[cfg(feature = "alloc")]
+    pub fn new() -> Processor<NullTimeProvider> {
+        Processor::new_with_time_provider(NullTimeProvider::default())
+    }
+
+    #[cfg(not(feature = "alloc"))]
+    pub fn new() -> Processor<NullTimeProvider> {
+        Processor {
+            parser: Parser::new(),
+            state: ProcessorState {
+                preceding_char: None,
+                _time_provider: PhantomData::<NullTimeProvider>::default(),
+            },
+        }
+    }
+}
+
+impl<T: TimeProvider> Processor<T> {
+    #[cfg(feature = "alloc")]
+    pub fn new_with_time_provider(provider: T) -> Processor<T> {
+        Processor {
+            parser: Parser::new(),
+            state: ProcessorState {
+                dcs: None,
+                preceding_char: None,
+                sync_state: SyncState::new(provider),
+                _time_provider: PhantomData::<T>::default(),
+            },
+        }
+    }
+
+    #[inline]
+    pub fn advance<H, W>(&mut self, handler: &mut H, byte: u8, writer: &mut W)
+    where
+        H: Handler<W>,
+    {
+        #[cfg(feature = "alloc")]
+        {
+            if !self.state.sync_state.in_sync {
+                let mut performer = Performer::new(&mut self.state, handler, writer);
+                self.parser.advance(&mut performer, byte);
+            } else {
+                self.advance_sync(handler, writer, byte)
+            }
+        }
+        #[cfg(not(feature = "alloc"))]
+        {
+            let mut performer = Performer::new(&mut self.state, handler, writer);
+            self.parser.advance(&mut performer, byte);
+        }
+    }
```

Comment:
What's the point of having separate blocks inside the method, when the entire method is different for each feature anyway?